### PR TITLE
[sso-cd] Stop using service ca from service account token

### PR DESCRIPTION
**Pull request by marun**
This change is required by [Bugzilla bug 1813894](https://bugzilla.redhat.com/show_bug.cgi?id=1813894). Inclusion of the service ca in token configmaps is discontinued in 4.5. This PR updates a template included by `openshift/library` which is in turn included in openshift releases by `openshift/cluster-samples-operator`. This change maintains template compatibility with previous openshift releases, and is required for compatibility with 4.5.

I'm marking WIP because I'm not sure selectively updating templates is acceptable. It's also not clear to me how to update the adoc for the template. A bare invocation of `gen_template_docs.py` returns the following:

```
Generating docs/templates/sso-cd-https.adoc...
Traceback (most recent call last):
  File "./gen_template_docs.py", line 354, in <module>
    generate_templates()
  File "./gen_template_docs.py", line 55, in generate_templates
    generate_template(os.path.join(directory, template))
  File "./gen_template_docs.py", line 79, in generate_template
    text_file.write(createTemplate(data, path))
  File "./gen_template_docs.py", line 159, in createTemplate
    return templater.render(tdata)
  File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/template.py", line 87, in render
    return self.formatter.format(template, **data)
  File "/usr/lib64/python3.7/string.py", line 186, in format
    return self.vformat(format_string, args, kwargs)
  File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 119, in vformat
    return self._vformat(string, args, kwargs)
  File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 128, in _vformat
    return self.formatsection(tokens, kwargs)
  File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 212, in formatsection
    result.append(self.format_field(obj, spec))
  File "/usr/lib64/python3.7/string.py", line 260, in format_field
    return format(value, format_spec)
TypeError: format() argument 2 must be str, not tuple
```

### CODE REVIEW

**Comment by marun** [comment]
cc @hmlnarik

**Comment by iankko** [comment]
@marun 

Hey Maru,

  thx for the heads-up! 

> This change is required by [Bugzilla bug 1813894](https://bugzilla.redhat.com/show_bug.cgi?id=1813894). Inclusion of the service ca in token configmaps is discontinued in 4.5. This PR updates a template included by `openshift/library` which is in turn included in openshift releases by `openshift/cluster-samples-operator`. This change maintains template compatibility with previous openshift releases, and is required for compatibility with 4.5.

Since the PR label mentions removing of the service CA from the service accoount, wanted to check if the change itself includes the [recommended](https://github.com/openshift/openshift-docs/blob/enterprise-4.1/release_notes/ocp-4-1-release-notes.adoc#service-ca-bundle-changes) `service.beta.openshift.io/inject-cabundle=true` annotation change, but after brief review (without testing) the change looks reasonable to me (seems to follow the aforementioned guidance).

> 
> I'm marking WIP because I'm not sure selectively updating templates is acceptable.

Assuming your change touched solely the `sso-cd-x509-https` template, because it's [the only one](https://github.com/openshift/library/blob/master/official.yaml#L1020) referenced in the `openshift/library` repo? But the type of change would be applicable to all provided / available `*-x509-*` templates, correct? If the latter is the case, would be great / preferable to get them updated / fixed in _all-in-one-batch_ PR / approach. But we can do that, if necessary..

 It's also not clear to me how to update the adoc for the template. A bare invocation of `gen_template_docs.py` returns the following:

You got it running correctly, just the script wasn't upgraded to work properly with Python 3 yet. So workaround is to change the shebang to `#!/usr/bin/env python2` temporarily, and generate the `*.adoc` files. No manual (re)generation of the `*.adoc` files is expected (we will fix the ` "./gen_template_docs.py"` script issue in the future - so far it wasn't urgent enough to be done immediately due to other priorities).

> 
> ```
> Generating docs/templates/sso-cd-https.adoc...
> Traceback (most recent call last):
>   File "./gen_template_docs.py", line 354, in <module>
>     generate_templates()
>   File "./gen_template_docs.py", line 55, in generate_templates
>     generate_template(os.path.join(directory, template))
>   File "./gen_template_docs.py", line 79, in generate_template
>     text_file.write(createTemplate(data, path))
>   File "./gen_template_docs.py", line 159, in createTemplate
>     return templater.render(tdata)
>   File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/template.py", line 87, in render
>     return self.formatter.format(template, **data)
>   File "/usr/lib64/python3.7/string.py", line 186, in format
>     return self.vformat(format_string, args, kwargs)
>   File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 119, in vformat
>     return self._vformat(string, args, kwargs)
>   File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 128, in _vformat
>     return self.formatsection(tokens, kwargs)
>   File "/home/maru/.local/lib/python3.7/site-packages/ptemplate/formatter.py", line 212, in formatsection
>     result.append(self.format_field(obj, spec))
>   File "/usr/lib64/python3.7/string.py", line 260, in format_field
>     return format(value, format_spec)
> TypeError: format() argument 2 must be str, not tuple
> ```

Yet, let me know, how do you want / plan to proceed further with this one (last but not least, besides the `sso-cd-x509-https` template update, the same / similar updates are necessary against the current RH-SSO 7.3 & 7.4 templates, included in `openshift/library` too, to get them continue working post the `4.6` service ca account removal / annotation change. So we would want to make those updates to all of them if possible).

HTH

Thanks!
Jan

**Comment by hmlnarik** [comment]
@marun Thanks for the heads-up! We'll need to treat the changes on a broader scope and to understand the scope precisely would like if you could answer @iankko's technical questions.

@iankko This needs to be addressed on Keycloak / RHSSO side by us. There are several items to it
- openshift library contains templates for CD builds which are long obsolete and need to be removed from there
- The update needs to be propagated to 7.4 templates
- 7.4 updated templates should go into openshift library

**Comment by marun** [comment]
@iankko 

> Assuming your change touched solely the `sso-cd-x509-https` template, because it's [the only one](https://github.com/openshift/library/blob/master/official.yaml#L1020) referenced in the `openshift/library` repo? But the type of change would be applicable to all provided / available `*-x509-*` templates, correct? If the latter is the case, would be great / preferable to get them updated / fixed in _all-in-one-batch_ PR / approach. But we can do that, if necessary..

I was attempting to update templates included in `openshift/cluster-samples-operator` (https://github.com/openshift/cluster-samples-operator/pull/270):

- sso72-x509-https.json
- sso73-x509-https.json
- sso73-x509-mysql-persistent.json
- sso73-x509-postgresql-persistent.json
- sso74-x509-https.json
- sso74-x509-postgresql-persistent.json

Then I realized that these templates were sourced by the operator from `openshift/library`, which in addition to the above templates also included `sso-cd-x509-https.json`. I decided to try to fix all templates in `openshift/library` to minimize the chances of someone using a template incompatible with ocp 4.5 and above, and this PR was the starting point.

If I understand the comments from you and @hmlnarik, the sso-cd and sso72 templates are outdated and can be removed from openshift/library rather than being updated. I've proposed 2 new PRs that update the sso73 and sso74 templates:

- https://github.com/jboss-container-images/redhat-sso-7-openshift-image/pull/145
- https://github.com/jboss-container-images/redhat-sso-7-openshift-image/pull/146

> You got it running correctly, just the script wasn't upgraded to work properly with Python 3 yet. So workaround is to change the shebang to `#!/usr/bin/env python2` temporarily, and generate the `*.adoc` files. No manual (re)generation of the `*.adoc` files is expected (we will fix the ` "./gen_template_docs.py"` script issue in the future - so far it wasn't urgent enough to be done immediately due to other priorities).

Switching to python2 worked, and all PRs now include updated docs.

> Yet, let me know, how do you want / plan to proceed further with this one (last but not least, besides the `sso-cd-x509-https` template update, the same / similar updates are necessary against the current RH-SSO 7.3 & 7.4 templates, included in `openshift/library` too, to get them continue working post the `4.6` service ca account removal / annotation change. So we would want to make those updates to all of them if possible).

I've updated this PR to fix all x509 templates in case its still useful. Feel free to close if not.

Your quick response is very much appreciated! Getting 1813894 resolved requires landing a lot of PRs across multiple repos before I can land the origin PR that removes the service ca bundle from token configmaps. The sooner that happens the better to leave as much time as possible to surface breakage before GA.

File: templates/sso-cd-x509-https.json [hunk 1]
```diff
@@ -369,6 +388,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/sso-cd-x509-https.json]: Replace `"configMapName"` with `"name"` like in RH-SSO 7.3 & 7.4 case

File: templates/sso-cd-x509-mysql-persistent.json [hunk 2]
```diff
@@ -492,6 +511,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/sso-cd-x509-mysql-persistent.json]: Yet here

File: templates/sso-cd-x509-postgresql-persistent.json [hunk 3]
```diff
@@ -474,6 +493,12 @@
                                 "secret": {
                                     "secretName": "sso-x509-jgroups-secret"
                                 }
+                            },
+                            {
+                                "name": "service-ca",
+                                "configMap": {
+                                    "configMapName": "${APPLICATION_NAME}-service-ca"
```
Comment (iankko) [inline templates/sso-cd-x509-postgresql-persistent.json]: And here

**Review by iankko — CHANGES_REQUESTED** [review]
Replace all instances of `"configMapName"` with `"name"`. The rest LGTM (verified the template works OK once done)

Thanks! Jan

**Comment by hmlnarik** [comment]
Thank you @marun 

@iankko I rely on you in this PR.

Yet as I mentioned above, I don't think fixing these templates is really necessary for CD (on the contrary 7.4), CD has been discontinued and templates should be removed from the library. For the sake of consistency,  this PR can be accepted as one last bit.

File: templates/sso-cd-x509-https.json (hunk 1 above)
Comment (marun) [reply templates/sso-cd-x509-https.json]: Done

File: templates/sso-cd-x509-mysql-persistent.json (hunk 2 above)
Comment (marun) [reply templates/sso-cd-x509-mysql-persistent.json]: Done

File: templates/sso-cd-x509-postgresql-persistent.json (hunk 3 above)
Comment (marun) [reply templates/sso-cd-x509-postgresql-persistent.json]: Done

**Comment by marun** [comment]
@iankko Updated, PTAL

**Comment by marun** [comment]
Not necessary

**Pull request closed without merging**